### ISSUE TITLE
Color images in PDFs are increasingly mangled by consecutive fullscreen refreshes

### ISSUE THREAD

**Issue by mr-scrawley** [koreader/koreader#11233]
* KOReader version: v2023.10-44-g3c63a5d8a_2023-12-04
* Device: Pocketbook InkPad Color 3 /  PBInkPadColor3 (PB743K3)

#### Issue

**When dithering is enabled, manually triggering a fullscreen refresh via gesture increasingly mangles the currently displayed color image in PDF documents (other document types untested).**
It looks like some processing step is executed recursively on the currently displayed content instead of the source image.

See the attached photos from https://freekidsbooks.org/wp-content/uploads/2019/12/The-Frog-Prince-Landscape-Book-CKF-FKB.pdf.
The one prefixed `dithering_off_` obviously has dithering disabled. Afterwards every photo has dithering enabled and every consecutive photo shows one more fullscreen refresh.

This effect is only visible with dithering set to on. Fullscreen refreshes with dithering off do not result in any perceived changes in color images.

If I understand #10446 correctly, dithering is currently not working on pocketbook devices at all. However, according to https://github.com/koreader/koreader/issues/10782#issuecomment-1780561662, color saturation modifications are only applied if the dithering toggle is set to "on", which does seem to be the case looking at the attached photos. Is this an accurate description of the current state of color rendering on pocketbook devices?

_BTW: Even with dithering set to off and the resulting somewhat muted colors, I actually prefer Koreader's result over the native pocketbook application. The latter produces excessively vibrant (or saturated?) images that especially degrade the appearance of real-life photos in my opinion._

#### Steps to reproduce

1. Open color pdf
2. Enable dithering
3. Trigger multiple fullscreen refreshes

##### `crash.log`

Verbose debug log is attached. I have tried to only execute relevant actions during the logged run.

#### Attachments

[PB743K3_dithering.log](https://github.com/koreader/koreader/files/13692668/PB743K3_dithering.log)

![20231216_133222](https://github.com/koreader/koreader/assets/17273004/d907219c-13a0-405e-8660-6d23bf57959f)
![20231216_133214](https://github.com/koreader/koreader/assets/17273004/6d20a1a5-e969-485d-909d-345d9393f87d)
![20231216_133209](https://github.com/koreader/koreader/assets/17273004/4026d804-c30f-4b0c-a972-a5b9e7b92d52)
![20231216_133204](https://github.com/koreader/koreader/assets/17273004/99b0f979-bdbf-418b-b459-273054a64c06)
![20231216_133159](https://github.com/koreader/koreader/assets/17273004/2e0e8adf-5aac-4efc-b905-a14ef5e7ab6c)
![20231216_133155](https://github.com/koreader/koreader/assets/17273004/2d17867f-f55a-4d96-9710-b87bcaaef4af)
![20231216_133150](https://github.com/koreader/koreader/assets/17273004/e433448d-da08-4ec1-942b-ee7565319966)
![20231216_133145](https://github.com/koreader/koreader/assets/17273004/170093ca-5df2-43e6-b405-0d4a23c943c3)
![dithering_off_20231216_133136](https://github.com/koreader/koreader/assets/17273004/1227b6e8-7922-4838-b2f6-059423c86ae7)

**Comment by NiLuJe** [comment]
I would assume the InkView processing happens on the working buffer, and the refresh gesture only requests a *refresh*, not a *repaint*.

Since said InkView processing is tied to refreshes, and not repaints, I imagine it's probably being repeated on an unchanged buffer (i.e., the processing gets stacked).

I don't think we really want to care about such a degenerate use-case?

**Comment by NiLuJe** [comment]
> If I understand #10446 correctly, dithering is currently not working on pocketbook devices at all. However, according to [#10782 (comment)](https://github.com/koreader/koreader/issues/10782#issuecomment-1780561662), color saturation modifications are only applied if the dithering toggle is set to "on", which does seem to be the case looking at the attached photos. Is this an accurate description of the current state of color rendering on pocketbook devices?

IIRC, yes.

**Comment by NiLuJe** [comment]
I don't recall if all the legacy diagonal swipes have been converted to Dispatcher, but if they have, or you are actually using a custom gesture, patching the Dispatcher variant to actually request a repaint during that gesture is trivial:

```diff
diff --git a/frontend/device/devicelistener.lua b/frontend/device/devicelistener.lua
index 2235fc052..527b8a65c 100644
--- a/frontend/device/devicelistener.lua
+++ b/frontend/device/devicelistener.lua
@@ -343,7 +343,7 @@ function DeviceListener:onFullRefresh()
     if self.ui and self.ui.view then
         self.ui:handleEvent(Event:new("UpdateFooter", self.ui.view.footer_visible))
     end
-    UIManager:setDirty(nil, "full")
+    UIManager:setDirty("all", "full")
 end
 
 -- On resume, make sure we restore Gestures handling in InputContainer, to avoid confusion for scatter-brained users ;).
```

**Comment by mr-scrawley** [comment]
> I don't recall if all the legacy diagonal swipes have been converted to Dispatcher, but if they have, or you are actually using a custom gesture, patching the Dispatcher variant to actually request a repaint during that gesture is trivial:
> 
> ```diff
> diff --git a/frontend/device/devicelistener.lua b/frontend/device/devicelistener.lua
> index 2235fc052..527b8a65c 100644
> --- a/frontend/device/devicelistener.lua
> +++ b/frontend/device/devicelistener.lua
> @@ -343,7 +343,7 @@ function DeviceListener:onFullRefresh()
>      if self.ui and self.ui.view then
>          self.ui:handleEvent(Event:new("UpdateFooter", self.ui.view.footer_visible))
>      end
> -    UIManager:setDirty(nil, "full")
> +    UIManager:setDirty("all", "full")
>  end
>  
>  -- On resume, make sure we restore Gestures handling in InputContainer, to avoid confusion for scatter-brained users ;).
> ```
> 
> (Should also be trivially adaptable to a user patch).

This does indeed fix the issue. Thank you so much!
(It works for the default short diagonal swipe as well as a custom gesture set to fullscreen refresh)

I assume by "degenerate use case" you mean image processing being delegated to InkView on Pocketbook devices? From what I understand so far, I can see why we wouldn't want to force a repaint on all devices during fullscreen refreshes. So this is indeed probably best implemented via user patch.

I can try to turn your patch into a user patch.

**Comment by mr-scrawley** [comment]
> I can try to turn your patch into a user patch.

This seems to work:
[2-repaint-on-fullscreen-refresh.lua.txt](https://github.com/koreader/koreader/files/13695051/2-repaint-on-fullscreen-refresh.lua.txt)
```lua
local DeviceListener = require("device/devicelistener")
local UIManager = require("ui/uimanager")

local onFullRefresh_orig = DeviceListener.onFullRefresh
DeviceListener.onFullRefresh = function(self)
    onFullRefresh_orig(self)
    UIManager:setDirty("all", "full")
end
```

Or is it not advisable to call the original function before, resulting in a redundant call of `UIManager:setDirty`?

**Comment by NiLuJe** [comment]
That function is trivial & small enough (and unlikely to change) that I'd just overload it directly with the full patched implementation ;).

But, to answer your actual question, since the original `setDirty` call doesn't flag anything for repaint, no, the redundancy is perfectly harmless here, so this works perfectly fine ;).

**Comment by NiLuJe** [comment]
> I assume by "degenerate use case" you mean

I meant the actual action of repeatedly full-refreshing the same page/screen with no movement or zooming being done ;).

**Comment by mr-scrawley** [comment]
> > I assume by "degenerate use case" you mean
> 
> I meant the actual action of repeatedly full-refreshing the same page/screen with no movement or zooming being done ;).

Ah, gotcha.
I guess it's mostly relevant for images and especially color images when rendered on current e-paper technology as there can be heavy ghosting even with flashing between pages.

At least that's the case in my experience. But I may just need to take another look at the flashing options. For now, refreshing multiple times has been my go-to fix to clean up images, which led me to discover this behavior in the first place.

**Comment by NiLuJe** [comment]
That you're specifically singling this out doesn't make my already middling interest in color eInk screen any better ;).

(It's definitely no longer an issue on modern high dpi grayscale screens).

**Comment by mr-scrawley** [comment]
> Slightly OT: That you're specifically singling this out doesn't make my already middling interest in color eInk screen any better ;).
> 
> (It's definitely no longer a noticeable issue on modern high dpi grayscale screens).

Small update: My flashing settings were indeed not configured ideally. I had assumed that the "always flash when page contains images" option would apply across the board, but I think it only affects EPUBs etc..
Setting the flash rate to every page vastly improves ghosting. There's still some ghosting but it's mostly acceptable and reduces the frequency with which I have to resort to manual refreshes.
The native pocketbook applications _seem_ to do even better in that regard, I wonder if they flash multiple times per turn or something.

**Comment by diegopau** [comment]
I am using an InkPad Color 3 and even using the "every page" + "always flash on pages with images", I still get more ghosting than with the default Pocketbook app on a CBZ file. In my case I always read manga and comic in landscape, so the pages are divided in 3 portions that I have to go through. Pocketbook app seems to fully get rid of the goshing each 2 taps maximum, sometimes I see zero ghosting even after each tap. So sometimes there is a bit of ghosting (even if it flashes) but the next tap will be a 100% clean from ghosting. It does seem to take a bit more time for each flash event but I don't notice multiple flashes (maybe it does so in a way that it looks like a longer continuous single flashing).
Using Koreader the ghosting can get pretty bad when reading for instance Manga, since some flat single color backgrounds many times are full of ghosting from the previous page or previous portion of the page. Everything else is handled better by Koreader though.

**Comment by mr-scrawley** [comment]
Awesome! Thanks for your contibutions!

I had already tested your changes in https://github.com/koreader/koreader-base/pull/1751.
Great to hear you have also found a fix these other issues.

I had already greatly enjoyed KOreader on my Inkpad before and am very happy to see so many of the quirks being ironed out :)

**Comment by Frenzie** [comment]
Included since a couple of weeks ago.

**Comment by mr-scrawley** [comment]
I think there has been a misunderstanding:
The improvements implemented by @dmalinovsky did not affect the problem originally described in this issue, but the general ghosting discussed later in the comments. @dmalinovsky never claimed as much, I must have missed that distinction in my general euphoria for these improvements.

I can confirm that the "mangling" on consecutive refresh ist still an issue without the [patch](https://github.com/koreader/koreader/issues/11233#issuecomment-1859058369). I just did not have much need for manual refreshes anymore after the other fixes 😅

The "issue with diagonal refresh" part of #11865 seems to be identical to what I have described in this issue.
